### PR TITLE
nrf_dm: Adding the nordic distancing lib

### PR DESCRIPTION
**Pull request by Tschet1**
The nrf_dm librarary is a proprietary distance measurement library that
allows to estimate the distance between two nordic devices that support
this feature.

Signed-off-by: Jan Müller <jan.mueller@nordicsemi.no>

Corresponding PR in nrf: https://github.com/nrfconnect/sdk-nrf/pull/6428

### CODE REVIEW

**Review by eriksandgren — APPROVED** [review]

File: nrf_dm/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+if(CONFIG_SOC_NRF52840)
+	set(NRF_DM_PATH ${CMAKE_CURRENT_SOURCE_DIR}/lib/pca10056/hard-float)
+elseif(CONFIG_SOC_NRF52833)
+	set(NRF_DM_PATH ${CMAKE_CURRENT_SOURCE_DIR}/lib/pca10100/hard-float)
+else()
+	message(FATAL_ERROR "Unsupported SoC.")
+endif()
```
Comment (rugeGerritsen) [inline nrf_dm/CMakeLists.txt:13]: You should consider unifying the path names to other libraries. This will also make it more future proof.
See the usage of `nrfxlib_calculate_lib_path` for other libs.

Comment (grochu) [reply nrf_dm/CMakeLists.txt:13]: Good point @rugeGerritsen. If you decide not to fully align the _nrf_dm_ paths at least please consider getting rid of the "_pca10xxx_" naming as Zephyr and NCS don't use these codes. E.g. Zephyr defines build target `nrf52dk_nrf52832` where `nrf52dk` refers to the name of the DK.

File: nrf_dm/Kconfig [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#
+# Copyright (c) 2020 Nordic Semiconductor
```
Comment (grochu) [inline nrf_dm/Kconfig:2]: Maybe 2021 for consistency with other files?

File: nrf_dm/README.rst [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+* The distance is computed based on all information available to the device.
+* All the mathematics necessary to compute distance is performed inside the library.
+
+Background
+==========
+
+Multi-carrier phase difference
+------------------------------
+
+Multi-carrier phase ranging for distance measurement is based on measuring the phase difference between two devices as a function of frequency.
```
Comment (grochu) [inline nrf_dm/README.rst:28]: Just a suggestion, does it sound better?
```suggestion
Multi-carrier phase ranging for distance measurement is based on measuring the phase difference in RF signals between two devices as a function of frequency.
```

File: nrf_dm/include/nrf_dm.h [hunk 4]
```diff
@@ -0,0 +1,310 @@
+/*
+ * Copyright (c) Nordic Semiconductor ASA
```
Comment (grochu) [inline nrf_dm/include/nrf_dm.h:2]: Please add year

File: nrf_dm/include/nrf_dm.h [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+		.ant_comp = NRF_DM_ANTENNA_COMP_1_1,                                               \
+		.dist_offset_err = { .ifft = 0.0f, .phase_slope = 0.0f, .rssi_openspace = 0.0f },  \
+	})
+
+/**
+ * @brief Initialize the distance measurement procedure
+ *
+ * @param[in] ppi_cfg PPI channels to be used by the library. The required amount of PPI channels
+ * is defined by the maximum hamming distance between two successive patterns in @ref
+ * switch_pattern divided by two On the nrf53 series, only 2 DPPI channels are used. A minimum of
```
Comment (grochu) [inline nrf_dm/include/nrf_dm.h:257]: Should it be:
```suggestion
 * switch_pattern divided by two. On the nrf53 series, only 2 DPPI channels are used. A minimum of
```

File: nrf_dm/license.txt [hunk 6]
```diff
@@ -0,0 +1,5 @@
+/*
+ * Copyright (c) Nordic Semiconductor ASA
```
Comment (grochu) [inline nrf_dm/license.txt:2]: Please add year

File: nrf_dm/include/nrf_dm_version.h [hunk 7]
```diff
@@ -0,0 +1,15 @@
+/*
+ * Copyright (c) Nordic Semiconductor ASA
```
Comment (grochu) [inline nrf_dm/include/nrf_dm_version.h:2]: Please add year

File: nrf_dm/README.rst [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+Overview
+********
+
+This library provides the following key features:
+
+* Easy-to-use interface for distance measurement.
+* The distance is computed based on all information available to the device.
+* All the mathematics necessary to compute distance is performed inside the library.
+
+Background
```
Comment (grochu) [inline nrf_dm/README.rst:22]: I am not familiar with the details of the ranging methods, but I've heard on some meetings that each method has its advantages in specific conditions: e.g. RTT might be more accurate for higher distances (I am not sure it's true, just an example). If any hint or characterization like this can be formulated, could you provide it here? It would help understand how to practically use this library.

**Review by grochu — COMMENTED** [review]
Added a few comments/suggestions.

**Comment by maje-emb** [comment]
Misspelling: `nrf_dm/lib/nNF52840/hard-float/libnrf_dm.a`?

File: nrf_dm/Kconfig [hunk 9]
```diff
@@ -0,0 +1,16 @@
+#
+# Copyright (c) 2021 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+config NRF_DM
+	bool "nRF DM (Distance Measurement) library [EXPERIMENTAL]"
+	select FPU
```
Comment (maje-emb) [inline nrf_dm/Kconfig:9]: Maybe?
```suggestion
	select FPU if CPU_HAS_FPU
```
To consider an SoC without an FPU

File: nrf_dm/include/nrf_dm.h [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+#endif
+
+/**
+ * @brief Quality of the procedure
+ */
+typedef enum {
+	NRF_DM_QUALITY_OK = 0,
+	NRF_DM_QUALITY_POOR = 1,
+	NRF_DM_QUALITY_DO_NOT_USE = 2,
+	NRF_DM_QUALITY_CRC_FAIL = 3,
```
Comment (maje-emb) [inline nrf_dm/include/nrf_dm.h:33]: Consider adding structure field comments, for consistency.

**Review by maje-emb — COMMENTED** [review]
Adding some comments.

File: nrf_dm/include/nrf_dm.h (hunk 10 above)
Comment (Tschet1) [reply nrf_dm/include/nrf_dm.h:33]: do you have a suggestion for the information to add?

Comment (maje-emb) [reply nrf_dm/include/nrf_dm.h:33]: Meybe:
```
/* Good measurement quality */
/* Poor measurement quality */
/* Measurement not for use */
```
I added this suggestion because almost all the structure fields in this file have comments.

Comment (Tschet1) [reply nrf_dm/include/nrf_dm.h:33]: Is this not already visible in the enum names?

Comment (maje-emb) [reply nrf_dm/include/nrf_dm.h:33]: Yes this is visible. It's just a suggestion.

Comment (Tschet1) [reply nrf_dm/include/nrf_dm.h:33]: ok, I will leave this as is, for now.

File: nrf_dm/README.rst (hunk 8 above)
Comment (Tschet1) [reply nrf_dm/README.rst:22]: We agree that such information could be useful. At this point, we don't have all the information needed to provide such a statement with enough confidence, however. We would probably add more details about how to use the different modes as we approach the production release.

Comment (grochu) [reply nrf_dm/README.rst:22]: Sounds good, thanks for clarifying.

**Review by grochu — APPROVED** [review]

**Review by FrancescoSer — APPROVED** [review]

File: nrf_dm/doc/nrf_dm_usage.rst [hunk 11]
```diff
@@ -9,82 +9,23 @@ Usage
...
-         float best;           /**< Best effort distance estimate */
-         } mcpd;
-
-         struct
-         {
-         float rtt;            /**< RTT: Distance estimate based on RTT measurement */
-         } rtt;
-      } nrf_dm_dist_estimates_t;
+#. Call the :c:func:`nrf_dm_report_get` function to get the results.
+   It returns a pointer to the :c:struct:`nrf_dm_report_t` structure which contains the the  :c:struct:`nrf_dm_dist_estimates_t` structure with an estimation of the distance.
```
Comment (FrancescoSer) [inline nrf_dm/doc/nrf_dm_usage.rst:31]: @Tschet1 Please fix the " the the  "

File: nrf_dm/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,20 @@
+#
+# Copyright (c) 2021 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+include(${NRFXLIB_DIR}/common.cmake)
```
Comment (tejlmand) [inline nrf_dm/CMakeLists.txt:7]: Why is this included here, when it's already loaded here:
https://github.com/nrfconnect/sdk-nrfxlib/blob/9869bf438ef973c68aaeac80c3431abb83b7caaa/CMakeLists.txt#L6

```suggestion
```

**Review by tejlmand — CHANGES_REQUESTED** [review]

**Review by FrancescoSer — APPROVED** [review]

**Comment by Tschet1** [comment]
rebased on current main

**Review by tejlmand — CHANGES_REQUESTED** [review]
why are the files: `libnrf_dm.a`
called libraries and with a `.a` extensions when they are not archive files but object files ?

I suggest to either add them directly as object files or create a proper archive containing just the single object file with proper name.

```
$ file libnrf_dm.a 
libnrf_dm.a: ELF 32-bit LSB relocatable, ARM, EABI5 version 1 (SYSV), not stripped
```

**Comment by Tschet1** [comment]
> why are the files: `libnrf_dm.a` called libraries and with a `.a` extensions when they are not archive files but object files ?
> 
> I suggest to either add them directly as object files or create a proper archive containing just the single object file with proper name.
> 
> ```
> $ file libnrf_dm.a 
> libnrf_dm.a: ELF 32-bit LSB relocatable, ARM, EABI5 version 1 (SYSV), not stripped
> ```

good point. fixed

**Comment by tejlmand** [comment]
> good point. fixed

Why is the extension of the file inside the archive having `.elf` ?
It is a relocatable object file, as one can see:
```
$ ar t nrf_dm/lib/nRF52840/hard-float/libnrf_dm.a
libnrf_dm.elf
$ ar x nrf_dm/lib/nRF52840/hard-float/libnrf_dm.a
$ file libnrf_dm.elf 
libnrf_dm.elf: ELF 32-bit LSB relocatable, ARM, EABI5 version 1 (SYSV), not stripped
```

and not an executable.
An executable has another signature, like this:
```
<name>: ELF 32-bit LSB executable, ARM, ....
```

so calling this file `.elf` is messing up terminology.

**Comment by Tschet1** [comment]
> so calling this file `.elf` is messing up terminology.
I see. My mistake.

@rugeGerritsen : I see that we are using ".elf" also for MPSL/SDC. Maybe we can update that, as well.

**Comment by rugeGerritsen** [comment]
> > so calling this file `.elf` is messing up terminology.
> > I see. My mistake.
> 
> @rugeGerritsen : I see that we are using ".elf" also for MPSL/SDC. Maybe we can update that, as well.

Yes we can

**Review by tejlmand — APPROVED** [review]
